### ISSUE TITLE
data_source_aws_instances.go hard coded for instance-state-name running

### ISSUE THREAD

**Issue by Gibby** [terraform-providers/terraform-provider-aws#3339]
### Affected Resource(s)
Please list the resources as a list, for example:
- data_source_aws_instances

We are trying to grab instance id's for instances that are not running for building a maintenance window task using data source of aws_instances, however instance-state-name is hardcoded and filtering on only running instances.

```
	params := &ec2.DescribeInstancesInput{
		Filters: []*ec2.Filter{
			&ec2.Filter{
				Name:   aws.String("instance-state-name"),
				Values: []*string{aws.String("running")},
			},
		},
	}
```

**Comment by radeksimko** [comment]
Hi @Gibby 
do you mind explaining the use case you have in mind?

> for building a maintenance window task

what kind of maintenance do you plan to build and how do you plan to perform it within the context of Terraform?
Generally speaking if the instance is not running most provisioners won't work.

Thanks.

**Comment by Gibby** [comment]
Hi @radeksimko 

We have some instances that are only powered on during the week when the developers need them. However we still power them on every Sunday to patch and perform maintenance. To build the maintenance window in terraform we had to hardcode the instance ID's instead of using a tag since the filter is hardcoded for running only. Having it hard coded for running does not follow the documentation which says you can use any filter from the AWS CLI reference.

**Comment by radeksimko** [comment]
I'm afraid I still don't fully understand how you perform the maintenance, where presumably the very first operation (as you mentioned) is turning the instance on.

We have some plans to possibly eventually support [changing state of the instance](https://github.com/terraform-providers/terraform-provider-aws/pull/3391) through Terraform, but it's a little bit more complex topic, so that is likely going to take some time. Regardless, this is not an option currently available in the provider, so my question is how are you turning these instances on?

Thanks.

**Comment by Gibby** [comment]
There are 2 ways we are starting instances for maintenance managed with our terraform templates. However, we have to hardcode the instance ID's in templates since we can not filter on non-running instances. We have maintenance windows in systems manager that we also manage with terraform templates.

Ways we start instances with terraform:
1) A lambda script that is triggered by a cloud even rule (managed with terraform)
2) Automation document in Systems Manager for starting instances (managed with terraform)

**Comment by radeksimko** [comment]
I see, so AWS SSM or Lambda does that trick for you, to either of which you'd like to pass IDs dynamically via interpolation in Terraform. That makes sense.

Thanks for all your patience in explaining the use case @Gibby , I think it's worth supporting this and we'd be happy to accept PR.

**Comment by bflad** [comment]
In version 1.26.0 of the AWS provider, releasing tonight or tomorrow, the `aws_instances` data source has been updated to optionally accept a new `instance_state_names` argument, which defaults to the previous behavior of just `running`, but also accepts other instance states like `stopped` as well.

**Comment by bflad** [comment]
This has been released in [version 1.26.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1260-july-04-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
